### ISSUE TITLE
Deno 1.13: Argument of type "jwk" is not assignable to parameter of type "raw"

### ISSUE THREAD

**Issue by seanaye** [invisal/god_crypto#39]
After upgrading to Deno v1.13.0 released today this library, specifically RSA can no longer be imported due to a typescript error.

in deno interpreter
```
> deno --version
1.13

> deno
> import { RSA }  from "https://deno.land/x/god_crypto/rsa.ts";
Uncaught TypeError: TS2345 [ERROR]: Argument of type '"jwk"' is not assignable to parameter of type '"raw"'.
    "jwk",
    ~~~~~
    at https://deno.land/x/god_crypto@v1.4.10/src/rsa/rsa_wc.ts:55:5
    at async <anonymous>:2:18

```

this works fine in Deno 1.12.x

**Comment by iugo** [comment]
Maybe it is related to <https://github.com/denoland/deno/issues/11481>. https://github.com/denoland/deno/pull/11662

**Comment by janusz-f** [comment]
I have the same error

**Comment by danopia** [comment]
The unfortunate thing is that this library doesn't need WebCrypto at all. It was written to automatically switch to subtle crypto when available. Which didn't go well when subtle crypto was added with none of the expected algorithms, and similarly limited types as well.

Perhaps a good route to restore this library's function would be removing the unsupported subtle crypto codepaths & restoring them once Deno ships a release with the needed features. I see `RSA-OAEP` and `AES-CBC` referenced by this library. Neither can be imported into Deno yet. (Deno 1.13.0 will not want to typecheck these statements at all, so more `@ts-ignore` would then be needed for backwards compat 😢)

It seems `/x/djwt` already jumped over to using WebCrypto directly, so for JWT purposes maybe `/x/jose` is all we'll get on Deno 1.13.

**Comment by lowlighter** [comment]
It should be fixed for next deno release though. It has been implemented on canary since from what I seen.

I don't remember the release cycle but I think it's 2-4 weeks for minors so eventually it'll autofix with time

**Comment by danopia** [comment]
Ok so there's two different types of problem resulting from Deno implementing Subtle Crypto.

The **first** is the typecheck issue described in this issue. Supposedly the fix https://github.com/denoland/deno/pull/11716 will be in the next Deno canary. It doesn't seem to be in the latest canary, but the merge was recent. So apps depending on signing things with external RSA keys will just need to skip the current Deno release. `--no-check` seems safe because the relevant code isn't called as of Deno 1.13.2.

The **second** issue is that `god_crypto` will use Subtle Crypto [when it's "supported"](https://github.com/invisal/god_crypto/blob/f7d131000a2fd24814414b084e519347ea191652/src/rsa/rsa_wc.ts#L77-L84). A recent Deno merge added enough APIs (https://github.com/denoland/deno/pull/11654) to trigger this. So on the latest canary `god_crypto` will now call Subtle Crypto for encryption/decryption operations, _which should fail_ until Deno's `importKey` learns RSA and/or AES. I don't see a public signal yet on when that would be.

In my previous comment I hadn't yet untangled the two different regressions. It's more clear now that the second issue would be a second issue. It hasn't triggered yet as of Deno 1.13.2 and the codepath is only used for encrypt/decrypt operations, so it's not a problem for JWTs & maybe the API will be ready by the next Deno minor. 🤷

**Comment by lowlighter** [comment]
@danopia Thanks a lot for the additional details 👍

**Comment by laughedelic** [comment]
Does the new release of Deno v1.14 help with fixing this?

**Comment by BayoKwendo** [comment]
@laughedelic actually it got rid of that crypto bugs. i see deno is growing fast

**Comment by danopia** [comment]
Yes, Deno v1.14 includes enough types for god_crypto to pass typecheck again. 

Also, as I warned above, Deno v1.14 now includes enough WebCrypto for god_crypto's **encryption/decryption** to fail. This is reproducible with the god_crypto test suite.

So for JWT stuff aka **signing/verifying** if you want to stick with god_crypto just skip Deno v1.13
